gluu-server/opt/gluu/jetty/oxauth/custom/scripts/application_session/AuthnInstant.py

```diff
@@ -15,21 +17,28 @@ class ApplicationSession(ApplicationSessionType):
     def __init__(self, currentTimeMillis):
         self.currentTimeMillis = currentTimeMillis
 
-    def init(self, configurationAttributes):
+    def init(self, customScript, configurationAttributes):
+        print ("%s: init" % customScript.getName())
         return True
 
     def destroy(self, configurationAttributes):
         return True
 
     def getApiVersion(self):
-        return 2
+        return 11
 
-    def startSession(self, httpRequest, sessionId, configurationAttributes):
-        sessionAttributes = sessionId.getSessionAttributes()
-        sessionAttributes.put(SessionIdService.SESSION_CUSTOM_STATE, Long.toString(sessionId.getExpirationDate().getTime()))
+    def startSession(self, httpRequest, session, configurationAttributes):
+        externalAuthenticationService = CdiUtil.bean(ExternalAuthenticationService)
+        sessionAttributes = session.getSessionAttributes()
+        # Exponse the session expiry timestamp to Passport via the session status API
+        sessionAttributes.put(SessionIdService.SESSION_CUSTOM_STATE, Long.toString(session.getExpirationDate().getTime()))
+
+        # If no MFA for this session, use the CSP's authentication instant as our own
+        sessionAcr = sessionAttributes.get("acr")
+        acrLevels = externalAuthenticationService.acrToLevelMapping()
         authnInstant = sessionAttributes.get("authnInstant")
-        if authnInstant:
-            sessionId.setAuthenticationTime(Date.from(Instant.parse(authnInstant)))
+        if acrLevels[sessionAcr] <= 50 and authnInstant:
```

Comment:
what the 50 mean? can we comment or use a static variable name

Reply:
Its checking the auth level of the current authentication script config to see if it includes MFA or not. All our MFA configs have an auth level > 50.

We can have it check the mfa_methods parameter instead to see if it is defined / non-empty. That's more readable however it would break the migration workflow in the "V2" script (not sure if that has a future though).